tests/unit/s2n_x509_validator_test.c

```diff
@@ -2188,5 +2189,130 @@ int main(int argc, char **argv)
+            uint8_t *chain_data = s2n_stuffer_raw_read(&cert_chain_stuffer, chain_len);
+            EXPECT_NOT_NULL(chain_data);
+
+            DEFER_CLEANUP(struct s2n_pkey public_key_out = { 0 }, s2n_pkey_free);
+            EXPECT_SUCCESS(s2n_pkey_zero_init(&public_key_out));
+            s2n_pkey_type pkey_type = S2N_PKEY_TYPE_UNKNOWN;
+            EXPECT_ERROR_WITH_ERRNO(s2n_x509_validator_validate_cert_chain(&validator, conn, chain_data, chain_len,
+                                            &pkey_type, &public_key_out),
+                    S2N_ERR_CERT_UNTRUSTED);
+            /* Failed while processing/reading in the cert chain */
+            EXPECT_TRUE(validator.state == INIT);
+        };
+
+        /* when the peer sends only compliant certs from a non-compliant CA, 
+         * validation fails on the local trust store 
+         */
+        {
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new_minimal(), s2n_config_ptr_free);
+            EXPECT_SUCCESS(s2n_config_add_pem_to_trust_store(config, (char *) &invalid_root_pem[0]));
+
+            DEFER_CLEANUP(struct s2n_x509_validator validator = { 0 }, s2n_x509_validator_wipe);
+            EXPECT_SUCCESS(s2n_x509_validator_init(&validator, &config->trust_store, 0));
+
+            DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(conn);
+            conn->security_policy_override = &security_policy_not_local;
+            EXPECT_SUCCESS(s2n_set_server_name(conn, "localhost"));
+
+            DEFER_CLEANUP(struct s2n_stuffer cert_chain_stuffer = { 0 }, s2n_stuffer_free);
+
+            /* `chain_pem_len - root_pem_len`: only use the first two certs in the chain (leaf and intermediate) to 
+             * ensure we are correctly failing on the local trust store and not the chain that the peer sent.
+             */
+            EXPECT_OK(s2n_test_cert_chain_data_from_pem_data(conn, &chain_pem[0], chain_pem_len - root_pem_len,
+                    &cert_chain_stuffer));
+            uint32_t chain_len = s2n_stuffer_data_available(&cert_chain_stuffer);
+            uint8_t *chain_data = s2n_stuffer_raw_read(&cert_chain_stuffer, chain_len);
+            EXPECT_NOT_NULL(chain_data);
+
+            DEFER_CLEANUP(struct s2n_pkey public_key_out = { 0 }, s2n_pkey_free);
+            EXPECT_SUCCESS(s2n_pkey_zero_init(&public_key_out));
+            s2n_pkey_type pkey_type = S2N_PKEY_TYPE_UNKNOWN;
+            EXPECT_ERROR_WITH_ERRNO(s2n_x509_validator_validate_cert_chain(&validator, conn, chain_data, chain_len,
+                                            &pkey_type, &public_key_out),
+                    S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_CERT);
+            /* X509_verify_cert finished successfully */
+            EXPECT_TRUE(validator.state == VALIDATED);
+        };
+
+        /* when certificate preference don't apply locally, certs in the trust store aren't validated */
```

Comment:
Meaning that we'll reject a chain with an invalid CA if the CA is sent by the peer, but not if we retrieve the same CA from the trust store ourselves? That doesn't seem like correct behavior. Shouldn't the "local" flag only apply to loading certificates, not to actual validation? So without it we should be able to configure an invalid cert chain to send to the peer, but it should have no effect on actually performing authentication.

Reply:
Discussed offline: changing this behavior. check_root_cert will always validate against cert key preferences, even if cert_prefs_apply_locally is false.